File: spec/factories/theme.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :theme do
    primary_color { 9_342_168 }
    primary_text { 16_777_215 }
    danger_color { 14_257_035 }
    danger_text { 16_777_215 }
    success_color { 12_573_067 }
    success_text { 16_777_215 }
    warning_color { 14_261_899 }
    warning_text { 16_777_215 }
    info_color { 9_165_273 }
    info_text { 16_777_215 }
    dark_color { 6_710_886 }
    dark_text { 15_658_734 }
    raised_background { 16_777_215 }
    background_color { 13_026_795 }
    body_text { 3_355_443 }
    muted_text { 3_355_443 }
    input_color { 15_789_556 }
    input_text { 6_710_886 }
    raised_accent { 16_250_871 }
    light_color { 16_316_922 }
    light_text { 0 }
  end
end

File: app/javascript/retrospring/utilities/theme.ts
export const THEME_MAPPING = {
  'primary_color': 'primary',
  'primary_text': 'primary-text',
  'danger_color': 'danger',
  'danger_text': 'danger-text',
  'warning_color': 'warning',
  'warning_text': 'warning-text',
  'info_color': 'info',
  'info_text': 'info-text',
  'success_color': 'success',
  'success_text': 'success-text',
  'dark_color': 'dark',
  'dark_text': 'dark-text',
  'light_color': 'light',
  'light_text': 'light-text',
  'raised_background': 'raised-bg',
  'raised_text': 'raised-text',
  'raised_accent': 'raised-accent',
  'raised_accent_text': 'raised-accent-text',
  'background_color': 'background',
  'body_text': 'body-text',
  'input_color': 'input-bg',
  'input_text': 'input-text',
  'input_placeholder': 'input-placeholder',
  'muted_text': 'muted-text'
};

export const getHexColorFromThemeValue = (themeValue: string): string => {
  return ('000000' + parseInt(themeValue).toString(16)).substr(-6, 6);
}

export const getDecimalTripletsFromHex = (hex: string): string => {
  return hex.match(/.{1,2}/g).map((value) => parseInt(value, 16)).join(', ');
}

export const getIntegerFromHexColor = (hex: string): number => {
  return parseInt(hex.substr(1, 6), 16);
}

export const getColorForKey = (key: string, color: string): string => {
  const hex = getHexColorFromThemeValue(color);

  if (key.includes('text') || key.includes('placeholder') || key.includes('rgb')) {
    return getDecimalTripletsFromHex(hex);
  } else {
    return `#${hex}`;
  }
};
